integration-cli/docker_cli_run_test.go

```diff
@@ -1203,6 +1203,38 @@ func (s *DockerCLIRunSuite) TestRunAddingOptionalDevicesInvalidMode(c *testing.T
 	}
 }
 
+func (s *DockerCLIRunSuite) TestRunAddingOptionalCdiDevice(c *testing.T) {
```

Comment:
It's preferred to add an `integration/` test, rather than adding an `integration-cli` test, as the latter is somewhat legacy. I would guess it could go in `integration/container` and look like one of the existing "run" tests, but with the desired `DeviceRequests` entry in the client API request.

That also makes the test independent of the cli version. I'm surprised this test works since the CLI-side of this hasn't been merged yet.

Reply:
Yes, I was wondering about this myself. I will have a look at adding an `integration/` test instead. Off the top of your head, do you have one that could be used as a starting point?

Reply:
I had a quick poke around, and for test setup, [`TestBuildSquashParent`](https://github.com/moby/moby/blob/90e82a2c058c98eb87e06c27ee058bd84cd55af0/integration/build/build_squash_test.go#L21) wouldn't be a bad place to start, since it's also testing an Experimental-flagged, Linux-only feature. I guess in this case, because you need to write the CDI config to the daemon's host's filesystem, then you won't be able to test with a remote daemon either.

I assume you also need to add the `"cdi"` feature to the config? The only integration test I noticed that manipulates is [`TestPingBuilderHeader`](https://github.com/moby/moby/blob/90e82a2c058c98eb87e06c27ee058bd84cd55af0/integration/system/ping_test.go#L124) to disable the on-by-default `"buildkit"` feature, so I guess that's an example for feature config. However, I didn't see the current test enabling this feature, so maybe I'm incorrectly reading the PR as being "default-off"?

For the actual test itself, it doesn't look like any existing integration tests work with DeviceRequests, so perhaps add a `WithDeviceRequest` helper to [`integration/internal/container/ops.go`](https://github.com/moby/moby/blob/90e82a2c058c98eb87e06c27ee058bd84cd55af0/integration/internal/container/ops.go#L213-L218), and then you're looking at something quite similar to the tests in [`integration/container/run_linux_test.go`](https://github.com/moby/moby/blob/90e82a2c058c98eb87e06c27ee058bd84cd55af0/integration/container/run_linux_test.go), e.g., [`TestPrivilegedHostDevices`](https://github.com/moby/moby/blob/90e82a2c058c98eb87e06c27ee058bd84cd55af0/integration/container/run_linux_test.go#L142), which also cannot run remotely as it needs to interact with the host daemon's filesystem.